### PR TITLE
Set `_fields` for return types for namedtuple check

### PR DESCRIPTION
**Pull request by XuehaiPan**
In PyTree utilities, we use `obj._fields` attribute to determine whether the object is a namedtuple:

https://github.com/pytorch/pytorch/blob/04082fc042d85ed934ccba4b74eb2fc1c2442aeb/torch/utils/_pytree.py#L89-L97

All members of `torch.return_types` are `PyStructSequence` types. As the CPython documented:

> Struct sequence objects are the C equivalent of `collections.namedtuple` objects.

However, the Python C API does not set the `_fields` attribute. This makes our namedtuple checker fail to test the `torch.return_types.*` as namedtuple. They need manually register as node type in PyTree utilities.

## BC-breaking notes

Everything in `torch.return_types` is a `PyStructSequence` type at the moment. Usually, an instance of that type would be considered a `namedtuple`, which is a common container type in the concept of pytrees. All `torch.return_types.*` types are implemented with the Python C API `PyStructSequence_InitType`, which is differ from the Python `collections.namedtuple`. Most of the PyTorch code detects the `namedtuple` by checking the `_fields` attribute. In this PR, we set the `_fields` attribute in the auto-generated type definition.

This changes the auto-generated C definition for the `return_types` types. That means C extensions built for previous versions of PyTorch need to be rebuilt to add the new class attribute `_fields`.

cc @ezyang @gchanan

### CODE REVIEW

File: tools/autograd/gen_python_functions.py [hunk 1]
```diff
@@ -712,10 +712,23 @@ def generate_return_type_definition_and_map_entry(
...
+        .name = "torch.return_types.{name}",
+        .doc = nullptr,
+        .fields = NamedTuple_fields,
+        .n_in_sequence = {len(fieldnames)}
+    }};
     if (!is_initialized) {{
         PyStructSequence_InitType(&{typename}, &desc);
         {typename}.tp_repr = (reprfunc)torch::utils::returned_structseq_repr;
+        // Set `_fields` attribute for namedtuple subclass check
+        PyObject* keys = PyTuple_New(desc.n_in_sequence);
```
Comment (albanD) [inline tools/autograd/gen_python_functions.py:725]:
```suggestion
        PyObject* keys = PyTuple_New(desc.n_in_sequence);
        if (!keys) {
          throw python_error();
        }
```

File: tools/autograd/gen_python_functions.py [hunk 2]
```diff
@@ -712,10 +712,23 @@ def generate_return_type_definition_and_map_entry(
...
+        for (Py_ssize_t i = 0; i < desc.n_in_sequence; ++i) {{
+            PyObject* key = PyUnicode_FromString(desc.fields[i].name);
```
Comment (albanD) [inline tools/autograd/gen_python_functions.py:727]: if (!key) {
          Py_DECREF(keys);
          throw python_error();
        }

File: tools/autograd/gen_python_functions.py [hunk 3]
```diff
@@ -712,10 +712,23 @@ def generate_return_type_definition_and_map_entry(
...
+            PyTuple_SET_ITEM(keys, i, key);
+        }}
+        PyObject_SetAttrString(reinterpret_cast<PyObject*>(&{typename}), "_fields", keys);
```
Comment (albanD) [inline tools/autograd/gen_python_functions.py:730]: You should check that the return code is not -1 here.

**Review by albanD — COMMENTED** [review]
Could you add testing for this please?
Also a small note explaining that this is BC-breaking and why in the PR description (for the release note).

**Comment by XuehaiPan** [comment]
Apart from the `_fields` attribute, there are some other semantic of `PyStructSequence` differ from `namedtuple`:

```python
inst = namedtuple_cls(*tup)
inst = structseq(tup)
```

**Pull request closed without merging**